File: sortColors.js
// Time Complexity : O(n) where n is the length of nums
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

// I am taking three pointers, low and mid will start from 0 and high will be the length of nums
var sortColors = function(nums) {
    let low = 0;
    let mid = 0;
    let high = nums.length-1;
    while(mid <= high) {
        // if mid is at 2 swap it with high and move high pointer by 1
        if(nums[mid] === 2) {
            let temp = nums[high];
            nums[high] = nums[mid];
            nums[mid] = temp;
            high--;
        // if mid is at 0 swap it with low and move mid and low pointer by 1
        } else if(nums[mid] === 0) {
            let temp = nums[low];
            nums[low] = nums[mid];
            nums[mid] = temp;
            mid++;
            low++;
        // if mid is 1 just move pointer by 1
        } else {
            mid++;
        }
    }
};

File: sortColors.go
// Time Complexity : O(n)
// Space Complexity : O(1) since it is sorted in place
// Did this code successfully run on Leetcode : Yes

/*

Approach:
We use 3 pointers (low, mid, high) for each of the colors. We move the mid
pointer only when we know that each element it encounters is sorted in the
right place. This approach also relies on the low and high pointers to be at
0th and n-1th positions respectively, as they will always point to the lowest
and highest parts of the array respectively.

*/

package main

func sortColors(nums []int) {
	if len(nums) == 0 {
		return
	}

	// consider 3 pointers for each of the three colors
	// low=0=red, mid=1=white, high=2=blue
	low, mid, high := 0, 0, len(nums)-1

	for mid <= high {
		if nums[mid] == 1 {
			mid++
		} else if nums[mid] == 2 {
			// swap mid and high
			temp := nums[mid]
			nums[mid] = nums[high]
			nums[high] = temp
			high--
		} else if nums[mid] == 0 {
			temp := nums[mid]
			nums[mid] = nums[low]
			nums[low] = temp
			low++
			mid++
		}
	}
}
